src/cls/lua/cls_lua.cc

```diff
@@ -0,0 +1,1057 @@
+
+  int ret = cls_cxx_remove(hctx);
+  return clslua_opresult(lua, (ret == 0), ret, 0);
+}
+
+/*
+ * cls_cxx_stat
+ */
+static int clslua_stat(lua_State *L)
+{
+  cls_method_context_t hctx = clslua_get_hctx(L);
+
+  uint64_t size;
+  time_t mtime;
+  int ret = cls_cxx_stat(hctx, &size, &mtime);
+  if (!ret) {
+    lua_pushinteger(L, size);
+    lua_pushinteger(L, mtime);
+  }
+  return clslua_opresult(L, (ret == 0), ret, 2);
+}
+
+/*
+ * cls_cxx_read
+ */
+static int clslua_read(lua_State *L)
+{
+  cls_method_context_t hctx = clslua_get_hctx(L);
+  int offset = luaL_checkinteger(L, 1);
+  int length = luaL_checkinteger(L, 2);
+  bufferlist *bl = clslua_pushbufferlist(L, NULL);
+  int ret = cls_cxx_read(hctx, offset, length, bl);
+  return clslua_opresult(L, (ret >= 0), ret, 1);
+}
+
+/*
+ * cls_cxx_write
+ */
+static int clslua_write(lua_State *L)
+{
+  cls_method_context_t hctx = clslua_get_hctx(L);
+  int offset = luaL_checkinteger(L, 1);
+  int length = luaL_checkinteger(L, 2);
+  bufferlist *bl = clslua_checkbufferlist(L, 3);
+  int ret = cls_cxx_write(hctx, offset, length, bl);
+  return clslua_opresult(L, (ret == 0), ret, 0);
+}
+
+/*
+ * cls_cxx_write_full
+ */
+static int clslua_write_full(lua_State *L)
+{
+  cls_method_context_t hctx = clslua_get_hctx(L);
+  bufferlist *bl = clslua_checkbufferlist(L, 1);
+  int ret = cls_cxx_write_full(hctx, bl);
+  return clslua_opresult(L, (ret == 0), ret, 0);
+}
+
+/*
+ * cls_cxx_getxattr
+ */
+static int clslua_getxattr(lua_State *L)
+{
+  cls_method_context_t hctx = clslua_get_hctx(L);
+  const char *name = luaL_checkstring(L, 1);
+  bufferlist  *bl = clslua_pushbufferlist(L, NULL);
+  int ret = cls_cxx_getxattr(hctx, name, bl);
+  return clslua_opresult(L, (ret >= 0), ret, 1);
+}
+
+/*
+ * cls_cxx_getxattrs
+ */
+static int clslua_getxattrs(lua_State *L)
+{
+  cls_method_context_t hctx = clslua_get_hctx(L);
+
+  map<string, bufferlist> attrs;
+  int ret = cls_cxx_getxattrs(hctx, &attrs);
+  if (ret < 0)
+    return clslua_opresult(L, 0, ret, 0);
+
+  lua_createtable(L, 0, attrs.size());
+
+  for (auto it = attrs.cbegin(); it != attrs.cend(); it++) {
+    lua_pushstring(L, it->first.c_str());
+    bufferlist  *bl = clslua_pushbufferlist(L, NULL);
+    *bl = it->second; // xfer ownership... will be GC'd
+    lua_settable(L, -3);
+  }
+
+  return clslua_opresult(L, 1, ret, 1);
+}
+
+/*
+ * cls_cxx_setxattr
+ */
+static int clslua_setxattr(lua_State *L)
+{
+  cls_method_context_t hctx = clslua_get_hctx(L);
+  const char *name = luaL_checkstring(L, 1);
+  bufferlist *bl = clslua_checkbufferlist(L, 2);
+  int ret = cls_cxx_setxattr(hctx, name, bl);
+  return clslua_opresult(L, (ret == 0), ret, 1);
+}
+
+/*
+ * cls_cxx_map_get_val
+ */
+static int clslua_map_get_val(lua_State *L)
+{
+  cls_method_context_t hctx = clslua_get_hctx(L);
+  const char *key = luaL_checkstring(L, 1);
+  bufferlist  *bl = clslua_pushbufferlist(L, NULL);
+  int ret = cls_cxx_map_get_val(hctx, key, bl);
```

Comment:
Do you mean negated to resolve to the correct errno string via strerror, or negated to change the logic?